### PR TITLE
Refactor and improve rendering of title/subtitle components in HTML/LaTeX tables

### PR DESCRIPTION
**Pull request by rich-iannone**
It would seem that the underlying bug that resulted in a LaTeX building error (when `subtitle` was `NULL` when calling `tab_header()`) has been solved (initial testing here confirms it's not a problem in `v0.3.0`). However, this PR is an opportunity to refactor the HTML and LaTeX rendering functions for the header component.

This guards against a LaTeX error that can occur if an empty string or a string entirely composed of spaces is provided in `subtitle`. The low-level `dt_heading_has_subtitle()` function has extra checks for these types of inputs. This also fixes HTML rendering when there's no subtitle since HTML markup with an empty string would be generated, resulting in extra vertical space upon rendering.

Several testthat snapshot tests were added.

Fixes: https://github.com/rstudio/gt/issues/463
Fixes: https://github.com/rstudio/gt/issues/623

### CODE REVIEW

File: R/dt_heading.R [hunk 1]
```diff
@@ -49,5 +49,9 @@ dt_heading_has_subtitle <- function(data) {
 
   heading <- dt_heading_get(data = data)
 
-  length(heading) > 0 && length(heading$subtitle) > 0 && !is.null(heading$subtitle)
+  length(heading) > 0 &&
+    length(heading$subtitle) > 0 &&
+    !is.null(heading$subtitle) &&
+    heading$subtitle != "" &&
+    !grepl("^\\s+$", heading$subtitle)
```
Comment (jcheng5) [inline R/dt_heading.R:56]:
```suggestion
    !grepl("\\S", heading$subtitle)
```

Comment (jcheng5) [inline R/dt_heading.R:56]:
```suggestion
    length(heading$subtitle) > 0 &&
    any(grepl("\\S", heading$subtitle))
```

File: R/utils_render_html.R [hunk 2]
```diff
@@ -239,163 +234,103 @@ create_heading_component <- function(data,
...
-      subtitle_classes <- c(subtitle_classes, "gt_bottom_border")
-    }
+  title_row <-
+    htmltools::tags$tr(
+      htmltools::tags$th(
+        colspan = n_cols,
+        class = paste(title_classes, collapse = " "),
+        style = title_styles,
+        htmltools::HTML(
+          heading$title %>% paste_right(footnote_title_marks)
```
Comment (jcheng5) [inline R/utils_render_html.R:306]:
```suggestion
         paste_right(heading$title, footnote_title_marks)
```

File: R/utils_render_html.R [hunk 3]
```diff
@@ -239,163 +234,103 @@ create_heading_component <- function(data,
...
       htmltools::tags$tr(
         htmltools::tags$th(
           colspan = n_cols,
-          class = paste(title_classes, collapse = " "),
-          style = title_styles,
+          class = paste(subtitle_classes, collapse = " "),
+          style = subtitle_styles,
           htmltools::HTML(
-            heading$title %>% paste_right(footnote_title_marks)
+            heading$subtitle %>% paste_right(footnote_subtitle_marks)
```
Comment (jcheng5) [inline R/utils_render_html.R:320]:
```suggestion
            paste0(heading$subtitle, footnote_subtitle_marks)
```

File: R/utils_render_latex.R [hunk 4]
```diff
@@ -70,6 +70,76 @@ create_table_start_l <- function(data) {
...
+      )
+
+    footnote_subtitle_marks <-
+      footnote_mark_to_latex(mark = footnote_subtitle_marks$fs_id_c)
+
+  } else {
+    footnote_subtitle_marks <- ""
+  }
+
+  title_row <-
```
Comment (jcheng5) [inline R/utils_render_latex.R:123]: For both `\large` and `\small`, use outer curly braces (maybe a helper function to create these). And remove the newline from the ends of title and subtitle; instead, use a newline at the beginning of subtitle (if there is a subtitle).

File: tests/testthat/test-table_parts.R [hunk 5]
```diff
@@ -83,6 +85,29 @@ test_that("a gt table contains the expected heading components", {
...
+    render_as_html() %>%
+    expect_snapshot()
+
+  # Expect that providing a subtitle value with a series
+  # a space characters also won't produce a subtitle line
+  mtcars_short %>%
+    gt() %>%
+    tab_header(title = "test title", subtitle = "   ") %>%
+    render_as_html() %>%
+    expect_snapshot()
```
Comment (jcheng5) [inline tests/testthat/test-table_parts.R:110]: Include a test case (for both HTML and LaTeX) with a non-empty subtitle

File: R/dt_heading.R [hunk 6]
```diff
@@ -49,5 +49,9 @@ dt_heading_has_subtitle <- function(data) {
 
   heading <- dt_heading_get(data = data)
 
-  length(heading) > 0 && length(heading$subtitle) > 0 && !is.null(heading$subtitle)
+  length(heading) > 0 &&
+    length(heading$subtitle) > 0 &&
```
Comment (jcheng5) [inline R/dt_heading.R:53]: It would be nice to have the logic for "non-zero-length, non-space-only string" as a utility function

File: R/dt_heading.R (hunk 1 above)
Comment (rich-iannone) [reply R/dt_heading.R:56]: I added the function `is_nonempty_string()` with `length(x) > 0 && any(grepl("\\S", x))` in the body. Now a majority of the `&&` chain is simplified here.

File: R/utils_render_html.R (hunk 2 above)
Comment (rich-iannone) [reply R/utils_render_html.R:306]: Changed.

File: R/utils_render_html.R (hunk 3 above)
Comment (rich-iannone) [reply R/utils_render_html.R:320]: Changed.

File: R/dt_heading.R (hunk 6 above)
Comment (rich-iannone) [reply R/dt_heading.R:53]: Now implemented. Thanks for the suggestion.

File: tests/testthat/test-table_parts.R (hunk 5 above)
Comment (rich-iannone) [reply tests/testthat/test-table_parts.R:117]: Now added.

File: R/utils_render_latex.R [hunk 7]
```diff
@@ -1,3 +1,7 @@
+# Create a simple LaTeX group by surrounding a statement with curly braces
+latex_group <- function(...) {
+  paste0("{", paste0(...), "}")
```
Comment (jcheng5) [inline R/utils_render_latex.R:3]:
```suggestion
  paste0("{", ..., "}")
```

**Review by jcheng5 — APPROVED** [review]
